satpy/writers/cf_writer.py

```diff
@@ -160,16 +164,121 @@ def make_time_bounds(dataarray, start_times, end_times):
 
 
+class AttributeEncoder(json.JSONEncoder):
+    """JSON encoder for dataset attributes"""
+    def default(self, obj):
+        """Returns a json-serializable object for 'obj'
+
+        In order to facilitate decoding, elements in dictionaries, lists/tuples and multi-dimensional arrays are
+        encoded recursively.
+        """
+        if isinstance(obj, dict):
+            serialized = {}
+            for key, val in obj.items():
+                serialized[key] = self.default(val)
+            return serialized
+        elif isinstance(obj, (list, tuple, np.ndarray)):
+            return [self.default(item) for item in obj]
+        return self._encode(obj)
+
+    def _encode(self, obj):
+        """Encode the given object as a json-serializable datatype.
+
+        Use the netcdf encoder as it covers most of the datatypes appearing in dataset attributes. If that fails,
+        return the string representation of the object."""
+        try:
+            return _encode_nc(obj)
+        except ValueError:
+            return str(obj)
+
+
+def _encode_nc(obj):
+    """Encode an arbitrary object in a netcdf compatible datatype
+
+    Raises:
+        ValueError if no netcdf compatible datatype could be found
+    """
+    if isinstance(obj, (bool, np.bool_)):
+        # Bool has to be checked first, because it is a subclass of int
+        return str(obj)
+    elif isinstance(obj, (int, float, str)):
+        return obj
+    elif isinstance(obj, np.integer):
+        return int(obj)
+    elif isinstance(obj, np.floating):
+        return float(obj)
+    elif isinstance(obj, np.void):
+        return tuple(obj)
+    elif isinstance(obj, np.ndarray):
+        if not len(obj.dtype) and obj.dtype == np.bool_:
+            # Convert array of booleans to array of strings
+            obj = obj.astype(str)
+        if not len(obj.dtype) and len(obj.shape) <= 1:
+            # Multi-dimensional nc attributes are not supported, so we have to skip record arrays and multi-dimensional
+            # arrays here
+            return obj.tolist()
+
+    raise ValueError('Unable to encode')
+
+
+def encode_nc(obj):
+    """Encode an arbitrary object in a netcdf compatible datatype
+
+    Try to find the best matching datatype. If that fails, encode as a string. Plain lists are encoded recursively.
+    """
+    if isinstance(obj, (list, tuple)) and all([not isinstance(item, (list, tuple)) for item in obj]):
+        return [encode_nc(item) for item in obj]
+    try:
+        return _encode_nc(obj)
+    except ValueError:
+        try:
+            # Decode byte-strings
+            decoded = obj.decode()
+        except AttributeError:
+            decoded = obj
+        return json.dumps(decoded, cls=AttributeEncoder).strip('"')
+
+
+def encode_attrs_nc(attrs):
+    """Encode dataset attributes in a netcdf compatible datatype
+
+    Args:
+        attrs (dict):
+            Attributes to be encoded
+    Returns:
```

Comment:
Does this render properly with sphinx or does there need to be a blank line before "Returns:"?

Reply:
Yes, it renders properly, no need for a blank line